Add formatting hints to numbers

wezm opened the pull request:
Fixes #328

Opening this PR for feedback, before proceeding further. The PR adds int hints so that number literals and number patterns retain their original format. For example:

```
$ echo 'let x : U32 = 1234; let _x : U16 = match x { 0x1234 => 1, 0b1111 => 2, "head" => 4,  _ => 3 }; _x' | mold -run cargo r -- elab
let x : U32 = 1234;
let _x : U16 = match x { 0b1111 => 2, 0x1234 => 1, "head" => 4, _ => 3 };
_x : U16
```

**Possible further work**

- ~~I think the extra `IntStyle` field in `Const` messes up `PartialEq`. Looks like I'll need to manually implement it and have it ignore that field.~~ — I've now done this
  - Oh you totally covered this in the issue I linked "One thing we'll need to consider is to ignore the style hint when comparing constants for equality. We could do this by a custom overload PartialEq, or, by implementing our own method (eg. Const::logical_eq), which will involve updates to surface::elaboration::unification and core::semantics"
- A way to add hints to binary data that is read, for example indicating that `table_id` in `table_record` in `opentype.fathom` should use the ASCII format.
- Pehaps find a better home for the traits etc. that were added to `core`.
- ~~Propagate the style from the inputs to ops~~ — I had a go at this
- ~~Add some more tests~~
- Allow hex/bin signed ints and extend this to them
- The ascii format uses big endian... should that be adjustable some how?

brendanzab commented on fathom/src/core.rs:
```diff
@@ -422,6 +431,88 @@ pub enum Const {
+impl IntStyle {
+    pub fn format<T: Styled<N>, const N: usize>(&self, number: T) -> String {
+        match self {
+            IntStyle::Binary => format! {"0b{:b}", number},
```
Why do we use braces for some of these `format!` invocations? Is that for some sort of escaping reason?

wezm replied:
I suspect I just mistyped `(` and being a macro it still works. It wasn't intentional, will fix.

brendanzab commented on fathom/src/core.rs:
```diff
@@ -422,6 +431,88 @@ pub enum Const {
+
+impl_styled_int!(u8, u16, u32, u64);
+
+pub trait Styled<const N: usize>:
```
Dunno if better name could be `IntStyled` or `UIntStyled`?

brendanzab commented on fathom/src/core.rs:
```diff
@@ -403,14 +403,23 @@ def_prims! {
-/// Constants
+/// Formatting style for integers
 #[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
+pub enum IntStyle {
```
Not sure if `UIntStyle` is a better name, given it is only used for unsigned integers. Not a big issue though - we can always defer this until we extend support to Signed integers as well.

brendanzab reviewed (commented):
Looking good so far! I really like the merging used in the primops - it's nice to try to preserve the styling where possible.

Good point about the endianness of the ASCII literals. I wasn't really sure about them when I first implemented them, but yeah, they might require some more thought. 🤔

wezm commented on fathom/src/core.rs:
```diff
@@ -422,6 +431,88 @@ pub enum Const {
+        }
+    }
+
+    pub fn merge(self, other: IntStyle) -> IntStyle {
+        use IntStyle::*;
+
+        // If they're the same then there's nothing to do
+        if self == other {
+            return self;
+        }
+
+        match (self, other) {
+            // If one is decimal then return the other
+            (Decimal, _) => other,
+            (_, Decimal) => self,
+            // otherwise favour ourself
+            (Hexadecimal, _) | (Binary, _) | (Ascii, _) => self,
+        }
+    }
```
> seems overly biased towards self

Absolutely. This was my lack of a better idea. Your suggestion is much better, thanks.

wezm commented on fathom/src/core/semantics.rs:
```diff
@@ -384,67 +390,67 @@ fn prim_step(prim: Prim) -> Option<PrimStep> {
-        Prim::U8And => const_step!([x: U8, y: U8] => Const::U8(u8::bitand(*x, *y))),
-        Prim::U8Or => const_step!([x: U8, y: U8] => Const::U8(u8::bitor(*x, *y))),
-        Prim::U8Xor => const_step!([x: U8, y: U8] => Const::U8(u8::bitxor(*x, *y))),
+        Prim::U8Add => const_step!([x, xst: U8, y, yst: U8] => Const::U8(u8::checked_add(*x, *y)?, xst.merge(*yst))),
```
I'm not fussed either way. I think I actually started with `IntStyle::merge` but happy to change it back.

wezm commented on fathom/src/core.rs:
```diff
@@ -403,14 +403,23 @@ def_prims! {
+    Binary,
+    Decimal,
+    Hexadecimal,
+    Ascii,
```
A little research suggests that the most common case is for these to be big-endian (E.g. [OSType](https://en.wikipedia.org/wiki/FourCC), PNG, AIFF, RIFF, AVI, WAV, DirectX, OpenType). Seems reasonable to just call it `Ascii` for now.

wezm commented:
Ok I addressed the comments. I originally opened this PR with some parts remaining but many have been addressed. So it's probably suitable for merging now, if you agree.

brendanzab reviewed (approved):
Looks good!
